rosbag2_transport/include/rosbag2_transport/rosbag2_transport.hpp

```diff
@@ -0,0 +1,90 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_TRANSPORT__ROSBAG2_TRANSPORT_HPP_
+#define ROSBAG2_TRANSPORT__ROSBAG2_TRANSPORT_HPP_
+
+#include <functional>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rosbag2/sequential_reader.hpp"
+#include "rosbag2/writer.hpp"
+#include "rosbag2_transport/play_options.hpp"
+#include "rosbag2_transport/record_options.hpp"
+#include "rosbag2_transport/storage_options.hpp"
+#include "rosbag2_transport/visibility_control.hpp"
+
+namespace rosbag2_transport
+{
+
+class GenericPublisher;
+class GenericSubscription;
+class Rosbag2Node;
+class Player;
+
+class Rosbag2Transport
+{
+public:
+  /// Default constructor
+  ROSBAG2_TRANSPORT_PUBLIC
+  Rosbag2Transport();
+
+  /// Constructor for testing, allows to set the reader and writer to use
+  ROSBAG2_TRANSPORT_PUBLIC
+  Rosbag2Transport(
+    std::shared_ptr<rosbag2::SequentialReader> reader, std::shared_ptr<rosbag2::Writer> writer);
+
+  ROSBAG2_TRANSPORT_PUBLIC
+  void init();
+
+  ROSBAG2_TRANSPORT_PUBLIC
+  void shutdown();
+
+  /**
+   * Records topics to a bagfile. Subscription happens at startup time, hence the topics must
+   * exist when "record" is called.
+   *
+   * \param storage_options Options regarding the storage (e.g. bag file name)
+   * \param record_options Options regarding the recording (e.g. the topics to record)
+   */
+  ROSBAG2_TRANSPORT_PUBLIC
+  void record(const StorageOptions & storage_options, const RecordOptions & record_options);
```

Comment:
I think I would prefer to have the topic here explicitly being passed via function arguments. I believe that makes the API clearer than hiding the topics in the record_options.

Reply:
I'm not completely sure I understand the merit of this. Having it the way it is currently means that the python part of the code only contains argument parsing and just dumps everything to `rosbag2_transport`, which is easier to test and work with. 

If the `rosbag2_transport` API is to be used by other tools, then I agree it might be a good idea to clarify which arguments are more important/essential, etc. But I have only a nebulous idea about the rosbag2_transport API at the moment. Maybe this could be postponed to when we have more information? We could open a ticket then.

Reply:
you can still do that, but having an overload function in this header file. I see it from the point that this header file can also be used by non-python functions and thus I believe it's easier to pass the topics explicitly as an array as well as having an overload which doesn't have any topics and thus subscribes to all of them.